### PR TITLE
Improved trajectory accuracy for all trenching commands

### PR DESCRIPTION
**Pull request by AstroStucky**
## Linked Issues:
Jira Ticket 🎟️ | [OCEANWATER-1131](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1131)

@anthonykoutroulis Please review code, testing is optional.
@kmdalal Please perform tests and review the design of each of the 3 trajectories. Code review is optional, but some of the contents of the `plan_trajectory` methods for the 3 actions may be helpful towards understanding trajectory design. 

## Summary of Changes
* Trajectories for **TaskGrind**, **TaskScoopCircular**, and **TaskScoopLinear** all rebuilt such that:
  * Commanded depths will be accurate assuming the provided ground_position (or z-coordinate) are accurate.
  * Consistent trajectory shape across the workspace. _e.g. The shape of a TaskScoopLinear at coordinate (1.7, 0.0) will be the same (1.7, 0.4)_
* A reference link, l_grinder_tip, added to the lander for accurate positioning of the grinder, similar to how its done for the scoop. 
* Updated defaults and some parameter descriptions in action client scripts.
* Updated the action client script of **TaskGrind** to match the format of other actions. The action definition remains the same. 
* Circular trajectory planning made possible in TrajectorySequence.

## Test 
Launch atacama_y1a so test results are consistent.
I tested resulting digging depths on the OCEANWATER-1069 branch, which already has code that computes scoop depth. The depth proved to be accurate within half a cm.  I don't expect the tester to go through the trouble of merging in another branch for testing purposes, so eyeballing the scoop depth should be just fine. Use the following image as reference. Note that the scoop is about 78 mm tall. 
Screenshot_20230503_125927

For each trench, we will interrogate the surface using **ArmFindSurface**, to achieve an accurate digging depth. 

We will start with **TaskGrind** and **TaskScoopLinear** at their default workspace position
1. Pick a spot along the trench to interrogate the surface height `rosrun arm_find_surface.py -x 1.65`
I get a height of -0.187, as seen in the last coordinate of the resulting log message, but yours may vary by a couple millimeters.
```
[INFO:/lander_action_servers] ArmFindSurface: Succeeded - ArmFindSurface trajectory stopped by a force of 315.17 N. Surface found at (1.660, 0.001, -0.187)
```
2. Now perform a grind using the acquired ground position `rosrun ow_lander task_grind.py -z -0.187`
3. Now dig `rosrun ow_lander task_scoop_linear.py -z -0.187`
Verify the trajectories look similar as in the video. Recording was made at a real-time factor of x3.

https://user-images.githubusercontent.com/17039973/236066336-e91b7c2e-2fb6-45c9-a047-8ce20f43589d.mp4

Now a perpendicular **TaskGrind** followed by a perpendicular **TaskScoopCircular**. We will assume the ground position is roughly the same.
4. `rosrun ow_lander task_grind.py -z -0.187 --perpendicular`
5. `rosrun ow_lander task_scoop_circular.py -x 1.45 -z -0.187 --perpendicular`
Verify the trajectories look  similar as in the video

https://user-images.githubusercontent.com/17039973/236066887-09a61c3b-5179-4770-9c76-cf86b71e18f5.mp4

## Freestyle Testing
I encourage reviewers to try different commands. There are many potential variations. 

## Caveat
You will notice it is difficult to select the correct (x, y) coordinate for a TaskScoop* that matches a TaskGrind if y =/= 0. This is because the (x, y) coordinate means something different in each command. For TaskGrind (parallel), (x, y) is halfway between the starting position and ending position of grind segments 1 and 2. For TaskGrind (perpendicular) (x, y) is halfway between each segment and in the center of each segment. For TaskScoopCircular (x, y) always represents the center of the trench, or the deepest part. For TaskScoopLinear (x, y) is the start of the linear segment; however, the entry point may be as much as 2 cm behind that point.

I have a proposal in mind to make the semantics of these commands more consistent and make a TaskGrind position easily translatable into a TaskScoop* position. However, I would like to get feedback on it from the rest of the team first, and I do not wish to inflate the scoop of this PR any further.

### CODE REVIEW

**Comment by kmdalal** [comment]
The tests passed, but there is a regression problem.  When I ran ReferenceMission1, the GuardedMove failed:
```
[ERROR]: Exception in your execute callback: 'RobotCommander' object has no attribute 'get_joints'
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/actionlib/simple_action_server.py", line 289, in executeLoop
    self.execute_callback(goal)
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/src/ow_lander/server.py", line 117, in __on_action_called
    self.execute_action(goal)
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/src/ow_lander/actions.py", line 171, in execute_action
    trajectory = self.plan_trajectory(goal)
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/src/ow_lander/actions.py", line 119, in plan_trajectory
    sequence = TrajectorySequence(
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/src/ow_lander/trajectory_sequence.py", line 29, in __init__
    self._joints_count = len(self._group.get_joints())
  File "/opt/ros/noetic/lib/python3/dist-packages/moveit_commander/robot.py", line 329, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'RobotCommander' object has no attribute 'get_joints'
```

**Comment by kmdalal** [comment]
I like the new, smooth, consistent trajectories.  My code review was light and insufficient, especially on the math parts (Terry may be best for these), but what I could follow looked reasonable.  I will retest once the regression problem I reported above is fixed; other than that I'm done with my review.

File: ow_lander/src/ow_lander/constants.py
Comment (AstroStucky) [reply ow_lander/src/ow_lander/constants.py:64]: Since you showed an interest in this, I thought I would share the following. In atacama_y1a the terrain height is about -0.18, and in europa_terminator_workspace it's about -0.16. So this is not really correct for either workspace. Perhaps this is just a default for something that should not be defaulted?

**Comment by AstroStucky** [comment]
@kmdalal The GuardedMove bug is fixed, but ReferenceMission1 still will not work because the grinds extend past the arm's max range. 

The reality is this PR probably breaks other plans in ow_plexil . I could draft up a sibling PR to fix those now, but since remaking PLEXIL plans is part of my next task, [OW-1102](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1102), I think I would prefer to save time and handle all PLEXIL plan updates in a single PR. 

Maybe this would be better merged into an intermediary epic branch for the time being so that noetic-devel doesn't break ow_plexil plans. What do you think?

It also just occurred to me that updates to ow_plexil plans should probably be held off at least until https://github.com/nasa/ow_autonomy/pull/114 is merged to avoid conflicts.

**Comment by mikedalal** [comment]
@AstroStucky I agree with both your suggestions: merge this into an intermediate/epic branch, and wait until my big autonomy PR is merged before doing the PLEXIL work.  In Dev Day I'd like to understand better how the arm's reach was affected.

File: ow_lander/src/ow_lander/constants.py
Comment (mikedalal) [reply ow_lander/src/ow_lander/constants.py:64]: Now that you mention it, I think it best not to default ground height.  It's been a "convenience" that is fine for our daily purposes, but conceptually should not be a default.

Here's an idea for an extension of this convenience, if feasible.  Maintain a mapping of our provided terrains (with default lander location) to some suitable ground height, if this can be easily looked up, by the arm actions as well as autonomy.

**Comment by AstroStucky** [comment]
I like this plan. It's not so much the arm's reach has been changed, it's more that (x, y) values mean something different for all 3 actions. I can discuss some simple changes to fix this in Tuesday's dev day.

**Comment by kmdalal** [comment]
GuardedMove works now, but the plan TestActions fails in all 3 of its ArmMoveCartesian/Guarded attempts.  Assuming this is due to the new X/Y semantics and won't be fixed for a while.  I'll approve this PR from my end as soon as the target is branch is changed.

File: ow_lander/src/ow_lander/constants.py
Comment (AstroStucky) [reply ow_lander/src/ow_lander/constants.py:64]: > Here's an idea for an extension of this convenience, if feasible. Maintain a mapping of our provided terrains (with default lander location) to some suitable ground height, if this can be easily looked up, by the arm actions as well as autonomy.

If the info was readily available to the lander, I would be all for this, but at the moment I do not know how to make lander actions aware of the name of the terrain.

Comment (AstroStucky) [reply ow_lander/src/ow_lander/constants.py:64]: The default values are also not a part of the action server because ROS Actions do not support default parameters. They instead only exist inside their corresponding action clients. So whatever operation it did to lookup the current terrain, it would have to do each call of an action client since those are not persistent.

**Comment by AstroStucky** [comment]
ArmMoveCartesian and GuardedMove failing come as a surprise to me. Those should have not been changed at all. I will investigate.

File: ow_lander/src/ow_lander/constants.py
Comment (mikedalal) [reply ow_lander/src/ow_lander/constants.py:64]: I want to retract my suggestion of maintaining maps to default ground locations because another modeling goal we should aim for is easily altering the placement of the lander on any terrain (Jira to come).  

I assume that a default ground height is useful only for argument-free running of the action client scripts.  It's never something that should be used in autonomy or a mission simulation.  To that end, hardcoding a default that works for our two most popular worlds seems fine for now, if we document that the default invocation may not work on other worlds.

**Comment by AstroStucky** [comment]
This now merges into the [OW-1146](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1146) epic branch to prevent PLEXIL plans from breaking before their parameter usage can be updated.

**Review by mikedalal — APPROVED** [review]

**Review by anthonykoutroulis — COMMENTED** [review]
Code looks good. I spent time on the math and can't see any issues, but my quaternion math knowledge is fairly limited.

**Comment by AstroStucky** [comment]
It turns out the failures in `TestActions.plx` are present in noetic-devel and are unrelated to this PR. They look to be relate to other parameter semantics that have changed, like the relative flag, with the exception of the GuardeMove bug mentioned earlier in this PR, and has been fixed. 

I can make sure the parameters in this plan make sense by the end of this epic branch this one merges into.

File: ow_lander/scripts/task_grind.py
Comment (AstroStucky) [reply ow_lander/scripts/task_grind.py:20]: We could. I don't feel very strongly about it one way or the other, but I think I would like it to be work in another PR.

**Review by anthonykoutroulis — APPROVED** [review]

**Review by kmdalal — APPROVED** [review]

**Pull request merged**
